src/server/src/main/java/io/cassandrareaper/core/StreamSession.java

```diff
@@ -0,0 +1,184 @@
+   *
+   * <p>The method assumes the stream has been previously initialized when {@link #addStream(String, SessionInfo, long)}
+   * was called. If it wasn't, the method does nothing.
+   *
+   * <p>If the notification comes with old timestamp, then no update happens.
+   *
+   * <p>This method is called every time Cassandra sends a stream update, which happens quite a lot.
+   *
+   * @param host originating the notification
+   * @param progressInfo the payload received from Cassandra with the JMX notification
+   * @param timeStamp of the JMX notification
+   * @return itself with the stream updated
+   */
+  public StreamSession updateStream(String host, ProgressInfo progressInfo, long timeStamp) {
+    String streamId = Stream.streamId(host, progressInfo.peer.toString());
+
+    streams.computeIfPresent(streamId, (id, oldStream) -> {
+      // process the payload only if the timestamp is newer
+      if (oldStream.getLastUpdated() < timeStamp) {
+        if (progressInfo.direction.equals(ProgressInfo.Direction.OUT)) {
+          oldStream.setSizeSent(progressInfo);
+        }
+        if (progressInfo.direction.equals(ProgressInfo.Direction.IN)) {
+          oldStream.setSizeReceived(progressInfo);
+        }
+        oldStream.setLastUpdated(timeStamp);
+      }
+      return oldStream;
+    });
+    return this;
+  }
+
+  /**
+   * Updates the stream to mark it as complete after Cassandra sends the related message.
+   *
+   * <p>The method assumes the stream has been initialized with {@link #addStream(String, SessionInfo, long)}.
+   * If it wasn't, then the method does nothing.
+   *
+   * <p>If the notification comes with old timestamps, then no update happens.
+   *
+   * @param host originating the notification
+   * @param peer the second node participating in the stream
+   * @param success of the streaming session between the two nodes
+   * @param timeStamp of the JMX notification
+   * @return itself with stream finalized
+   */
+  public StreamSession finalizeStream(String host, String peer, Boolean success, long timeStamp) {
+    String streamId = Stream.streamId(host, peer);
+    streams.computeIfPresent(streamId, (id, oldStream) -> {
+      if (oldStream.getLastUpdated() < timeStamp) {
+        oldStream.setCompleted(true);
+        oldStream.setSuccess(success);
+      }
+      oldStream.setLastUpdated(timeStamp);
+      return oldStream;
+    });
+    return this;
+  }
+
+  @JsonPOJOBuilder(buildMethodName = "build", withPrefix = "with")
+  public static final class Builder {
+
+    private String planId;
+    private Map<String, Stream> streams;
+
+    private Builder() {
+    }
+
+    public Builder withPlanId(String planId) {
+      this.planId = planId;
+      return this;
+    }
+
+    public Builder withNoStreams() {
+      this.streams = Maps.newConcurrentMap();
+      return this;
+    }
+
+    public StreamSession build() {
+      return new StreamSession(this);
```

Comment:
Both `planId` and `streams` could be null.
I don't think you'd want that. And if so, guava precondition checks should be in place here.